Use HttpCompletionOption.ResponseHeadersRead to reduce allocations

juanrodriguezcen opened the pull request:
`HttpClient.SendAsync` by default uses `HttpCompletionOption.ResponseContentRead` which buffers the entire response into a `MemoryStream` (creating a copy) before returning. 

`HttpCompletionOption.ResponseHeadersRead` returns control after the headers are read and then if the stream is needed, whenever `HttpResponseMessage.Content.ReadAsStreamAsync()` is called, the network stream is returned instead of the in-memory buffered copy. 

I did some benchmarks before and after using `HttpCompletionOption.ResponseContentRead` in the different `Expect` extension methods. These were the results:

Before:

![image](https://user-images.githubusercontent.com/6766854/88821617-479e9c80-d199-11ea-9639-97ac6905398b.png)

After:

![image](https://user-images.githubusercontent.com/6766854/88822010-b845b900-d199-11ea-969e-7c443c9d1e52.png)


Note: All requests were done to this url: "https://jsonplaceholder.typicode.com/comments"

The biggest impact was on `ExpectHttpSuccess` and on `ExpectJson`, so this PR proposes using `HttpCompletionOption.ResponseContentRead` there. I wasn't able to test `ExpectProtobuf`, but it's possible that it could benefit of it as well.

m0sa commented:
We have to be especially careful in exception scenarios here. We might have some places where we assume the response is buffered after deserialization fails, and we want to get the full raw body content for the exception.

deanward81 commented:
Good point @m0sa - I wonder if we need to have a way to revert the behaviour so we can support those diagnostic scenarios? Maybe something like `UseBuffering` (like .NET Core does) that puts completion back to `HttpCompletionOption.ResponseContentRead`?

juanrodriguezcen commented:
@m0sa The `HttpResponseMessage` [is disposed](https://github.com/StackExchange/StackExchange.Utils/blob/5e0b73805e2710e28b3350773049217fd6512ce6/src/StackExchange.Utils.Http/Http.cs#L61) early on, which disposes the content stream. The exception that is created inside that using already doesn't read the content, and after the user the content stream is not readable. In the application code doing `await response.RawResponse.Content.ReadAsString()` fails, even with `HttpCompletionOption.ResponseContentRead`.

Maybe I'm confused about the scenario that you are describing, would you mind to elaborate more about it?

m0sa commented:
It is disposed, but it's used downstream, with the assumption that the Content is there, e.g.:

https://github.com/StackExchange/StackExchange.Utils/blob/5e0b73805e2710e28b3350773049217fd6512ce6/src/StackExchange.Utils.Http/Http.cs#L91

... which uses response.Content ->

https://github.com/StackExchange/StackExchange.Utils/blob/5e0b73805e2710e28b3350773049217fd6512ce6/src/StackExchange.Utils.Http/HttpCallResponse.cs#L105


There could be other cases, in the handlers -> https://github.com/StackExchange/StackExchange.Utils/blob/5e0b73805e2710e28b3350773049217fd6512ce6/src/StackExchange.Utils.Http/Http.cs#L72

We also expose the raw response, so there could be more places (outside of this lib) relying on the current behavior https://github.com/StackExchange/StackExchange.Utils/blob/5e0b73805e2710e28b3350773049217fd6512ce6/src/StackExchange.Utils.Http/HttpCallResponse.cs#L30

juanrodriguezcen commented:
@m0sa I think it would be ok. The response not being buffered doesn't mean that it won't be accessible (as long a the `HttpResponseMessage`  has not been disposed yet), `responseMessage.Content.ReadAsStreamAsync()` can still be called, the main difference is that the stream won't be seekable (because it's the network stream itself). 

In `HttpCallResponse` only headers are used. Also at the place where  `HttpCallResponse` is created, the stream wouldn't be accessible because the `HttpResponseMessage` has already been disposed.

Regarding the handlers, that shouldn't be affected, because `HttpCompletionOption.ResponseHeadersRead` would only be set on the `ExpectXXXX` extension methods, and a `.WithHandler` can't be chained after an `ExpectXXXX`. 

About exposing `RawResponse`, the behavior wouldn't change. Even if the response body was buffered by the usage of `HttpCompletionOption.ResponseContentRead`, that `MemoryStream` has been disposed and is not accessible by the time that `HttpCallResponse.RawResponse` is returned.

juanrodriguezcen commented:
@deanward81 @NickCraver @m0sa I added the extension method `WithoutResponseBuffering` so that the consumer of the library can opt-in/out of buffering depending on their needs. 

Additionally I added the benchmarks project. These are the results of a new run, the previous one was only for netcoreapp31, this one includes .net472 as well:

|                                Method |           Job |       Runtime |     Mean |   Error |  StdDev |     Gen 0 |     Gen 1 |     Gen 2 |  Allocated |
|-------------------------------------- |-------------- |-------------- |---------:|--------:|--------:|----------:|----------:|----------:|-----------:|
|    'ExpectHttpSuccess with buffering' |    .NET 4.7.2 |    .NET 4.7.2 | 140.1 ms | 2.74 ms | 4.02 ms |         - |         - |         - | 4208.35 KB |
| 'ExpectHttpSuccess without buffering' |    .NET 4.7.2 |    .NET 4.7.2 | 430.0 ms | 8.30 ms | 7.36 ms |         - |         - |         - |   96.05 KB |
|           'ExpectJson with buffering' |    .NET 4.7.2 |    .NET 4.7.2 | 145.3 ms | 2.82 ms | 3.13 ms |         - |         - |         - | 6128.54 KB |
|        'ExpectJson without buffering' |    .NET 4.7.2 |    .NET 4.7.2 | 143.3 ms | 2.77 ms | 2.96 ms |         - |         - |         - | 1984.23 KB |
|         'ExpectString with buffering' |    .NET 4.7.2 |    .NET 4.7.2 | 142.4 ms | 2.79 ms | 3.72 ms |         - |         - |         - | 8513.51 KB |
|      'ExpectString without buffering' |    .NET 4.7.2 |    .NET 4.7.2 | 140.8 ms | 2.68 ms | 2.63 ms |         - |         - |         - |  4681.2 KB |
|      'ExpectByteArray with buffering' |    .NET 4.7.2 |    .NET 4.7.2 | 139.2 ms | 2.67 ms | 2.62 ms |         - |         - |         - | 5254.73 KB |
|   'ExpectByteArray without buffering' |    .NET 4.7.2 |    .NET 4.7.2 | 143.6 ms | 2.61 ms | 2.44 ms |         - |         - |         - | 5262.73 KB |
|    'ExpectHttpSuccess with buffering' | .NET Core 3.1 | .NET Core 3.1 | 142.5 ms | 2.74 ms | 3.84 ms |         - |         - |         - | 2944.56 KB |
| 'ExpectHttpSuccess without buffering' | .NET Core 3.1 | .NET Core 3.1 | 108.9 ms | 1.99 ms | 2.92 ms |         - |         - |         - |   16.29 KB |
|           'ExpectJson with buffering' | .NET Core 3.1 | .NET Core 3.1 | 146.4 ms | 2.73 ms | 3.55 ms |         - |         - |         - | 5162.77 KB |
|        'ExpectJson without buffering' | .NET Core 3.1 | .NET Core 3.1 | 140.8 ms | 2.51 ms | 2.89 ms |         - |         - |         - | 1900.81 KB |
|         'ExpectString with buffering' | .NET Core 3.1 | .NET Core 3.1 | 157.1 ms | 2.89 ms | 5.71 ms | 1000.0000 | 1000.0000 | 1000.0000 | 7473.81 KB |
|      'ExpectString without buffering' | .NET Core 3.1 | .NET Core 3.1 | 143.4 ms | 2.81 ms | 4.29 ms |         - |         - |         - | 4217.74 KB |
|      'ExpectByteArray with buffering' | .NET Core 3.1 | .NET Core 3.1 | 139.7 ms | 2.73 ms | 2.80 ms | 1000.0000 | 1000.0000 | 1000.0000 | 4321.21 KB |
|   'ExpectByteArray without buffering' | .NET Core 3.1 | .NET Core 3.1 | 146.5 ms | 2.91 ms | 4.78 ms |         - |         - |         - | 4318.37 KB |

NickCraver commented on Benchmarks/Benchmarks.csproj:
```diff
@@ -0,0 +1,15 @@
+  <ItemGroup>
+    <ProjectReference Include="..\src\StackExchange.Utils.Http\StackExchange.Utils.Http.csproj" />
+  </ItemGroup>
+</Project>
```
This change looks good, but we should nest this per conventions, e.g.:
```
benchmarks/StackExchange.Utils.Http.Benchmarks/StackExchange.Utils.Http.Benchmarks.csproj
```
Could do this after merge if no time allows - just seeing this was never merged today, woops.

NickCraver replied:
@juanrodriguezcen Did you see this one? I can relocate this post-merge too, just didn't want to step on toes.

juanrodriguezcen replied:
@NickCraver oops, I missed it, I added a new commit that moves and renames the benchmark project. Feel free to move/rename it if I didn't use the convention correctly.

NickCraver commented on Benchmarks/Program.cs:
```diff
@@ -0,0 +1,13 @@
+{ 
+    public class Program
+    {
+        public static void Main(string[] args)
+        {
+             BenchmarkRunner.Run<ResponseBuffering>();
+        }
```
Let's make this runnable for any future as well:
```suggestion
        public static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);
```

NickCraver commented on src/StackExchange.Utils.Http/Extensions.Modifier.cs:
```diff
@@ -201,5 +201,16 @@ public static IRequestBuilder WithProtocolVersion(this IRequestBuilder builder,
+        /// <summary>
+        /// Indicates that the response's content shouldn't be buffered, setting the HttpCompletionOption accordingly.
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <returns></returns>
```
Let's remove if not populated on the public docs please!
```suggestion
```

NickCraver commented on src/StackExchange.Utils.Http/HttpBuilder.cs:
```diff
@@ -15,6 +15,7 @@ internal class HttpBuilder : IRequestBuilder
         public IEnumerable<HttpStatusCode> IgnoredResponseStatuses { get; set; } = Enumerable.Empty<HttpStatusCode>();
         public TimeSpan Timeout { get; set; }
         public IWebProxy Proxy { get; set; }
+        public bool BufferResponse { get; set; }
```
```suggestion
        public bool BufferResponse { get; set; } = true;
```

NickCraver commented on src/StackExchange.Utils.Http/HttpBuilder.cs:
```diff
@@ -28,6 +29,7 @@ public HttpBuilder(string uri, HttpSettings settings, string callerName, string
             };
             Settings = settings;
             Timeout = (settings ?? Http.DefaultSettings).DefaultTimeout;
+            BufferResponse = true;
```
In a serialization context it's possible to skip constructors - let's initialize like `LogErrors` consistently above please :)
```suggestion
```

NickCraver commented on src/StackExchange.Utils.Http/IRequestBuilder.cs:
```diff
@@ -41,6 +41,12 @@ public interface IRequestBuilder
         TimeSpan Timeout { get; set; }
 
+        /// <summary>
+        /// Indicate if the response content should be buffered. Sets the HttpCompletionOption to use on this request accordingly.
```
```suggestion
        /// Indicate if the response content should be buffered (e.g. for access later, vs. only streamed when false). 
        /// Sets the HttpCompletionOption to use on this request accordingly.
```

NickCraver reviewed (changes requested):
Some changes inline - happy to take and do these especially since we dropped the ball on this one, just shout either way please - not trying to step on toes!

NickCraver reviewed (approved):
Looking good! Thanks for the tweaks <3
